### PR TITLE
many: container validation improvements

### PR DESCRIPTION
**Pull request by ZeyadYasser**

### CODE REVIEW

**Review by pedronis — APPROVED** [review]
thanks

**Review by ernestl — APPROVED** [review]
Thanks

**Comment by alexmurray** [comment]
LGTM!

File: snap/container.go [hunk 1]
```diff
@@ -78,6 +85,142 @@ var (
...
+		// don't follow absolute targets
+		if filepath.IsAbs(target) {
+			return symlinkInfo{target, os.FileMode(0), naiveTarget, true}, nil
+		}
+
+		// evaluate target relative to symlink directory
+		target = filepath.Join(filepath.Dir(path), target)
+
+		// target escapes container, cannot evaluate further, let's return
+		if strings.Split(target, string(os.PathSeparator))[0] == ".." {
```
Comment (zyga) [inline snap/container.go:166]: This is somewhat costly for what is requested. Is this somehow different than `strings.HasPrefix` or am I missing something subtle?

File: snap/container.go [hunk 2]
```diff
@@ -78,6 +85,142 @@ var (
...
+		path = target
+	}
+
+	return symlinkInfo{}, fmt.Errorf("too many levels of symbolic links")
+}
+
+func shouldValidateSymlink(path string) bool {
+	// we only check meta directory for now
+	pathTokens := strings.Split(path, string(os.PathSeparator))
+	if pathTokens[0] == "meta" {
```
Comment (zyga) [inline snap/container.go:191]: Is this `strings.HasPrefix`?

File: snap/container.go [hunk 3]
```diff
@@ -78,6 +85,142 @@ var (
...
+		return true
+	}
+	return false
+}
+
+func evalAndValidateSymlink(c Container, path string) (symlinkInfo, error) {
+	pathTokens := strings.Split(path, string(os.PathSeparator))
```
Comment (zyga) [inline snap/container.go:198]: Is this `strings.HasPrefix` again?

File: snap/container.go [hunk 4]
```diff
@@ -78,6 +85,142 @@ var (
...
+	info, err := evalSymlink(c, path)
+	if err != nil {
+		return symlinkInfo{}, err
+	}
+
+	if info.isExternal {
+		return symlinkInfo{}, fmt.Errorf("external symlink found: %s -> %s", path, info.naiveTarget)
+	}
+
+	// symlinks like this don't look innocent
```
Comment (zyga) [inline snap/container.go:213]: Can you clarify what is innocent here?

File: snap/container.go [hunk 5]
```diff
@@ -196,20 +339,32 @@ func validateContainer(c Container, needsrx, needsx, needsr, needsf, noskipd map
 			return nil
 		}
 
-		if needsrx[path] || mode.IsDir() {
+		if mode&os.ModeSymlink != 0 && shouldValidateSymlink(path) {
+			symlinkInfo, err := evalAndValidateSymlink(c, path)
+			if err != nil {
+				logf("%s", err)
```
Comment (zyga) [inline snap/container.go:345]: I would add a context here, like `"Cannot evaluate and validate symbolic link: %s"`

**Review by zyga — COMMENTED** [review]
I didn't read test code but I have some comments on the evaluation details. Have a look please.

**Pull request merged**